pkg/segment/writer/metrics/wal/wal.go

```diff
@@ -0,0 +1,297 @@
+}
+
+func (it *WalIterator) Next() (*WALDatapoint, bool, error) {
+	if it.currentIndex < len(it.currentBlock) {
+		dp := &it.currentBlock[it.currentIndex]
+		it.currentIndex++
+		return dp, true, nil
+	}
+
+	var blockSize uint32
+	err := binary.Read(it.file, binary.LittleEndian, &blockSize)
+	if errors.Is(err, io.EOF) {
+		log.Infof("Next: reached end of WAL file")
+		return nil, false, nil
+	} else if err != nil {
+		log.Errorf("Next: failed to read block size from WAL file: %v", err)
+		return nil, false, err
+	}
+
+	blockData := make([]byte, blockSize)
+	_, err = io.ReadFull(it.file, blockData)
+	if err != nil {
+		log.Errorf("Next: failed to read block data of size %d: %v", blockSize, err)
+		return nil, false, err
+	}
+
+	blockBuf := bytes.NewReader(blockData)
+	newBlock, err := decodeWALBlock(blockBuf)
+
+	if err != nil {
+		log.Errorf("Next: dataDecompression failed: %v", err)
+		return nil, false, err
+	}
+
+	it.currentBlock = newBlock
+	it.currentIndex = 1
+
+	return &newBlock[0], true, nil
+}
+
+func (it *WalIterator) Close() error {
+	return it.file.Close()
+}
+
+func encodeWALBlock(dps []WALDatapoint, compressedBlockBuf *bytes.Buffer) error {
+	N := uint32(len(dps))
+	if N == 0 {
+		log.Warn("EncodeWALBlock: received empty data points slice")
+		return errors.New("empty data points")
+	}
+
+	rawBlockBuf := &bytes.Buffer{}
+
+	err := binary.Write(rawBlockBuf, binary.LittleEndian, N)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write datapoint count: %v", err)
+		return err
+	}
+
+	timestamps := make([]uint64, N)
+	dpVals := make([]float64, N)
+	tsids := make([]uint64, N)
+	for i, dp := range dps {
+		timestamps[i] = dp.Timestamp
+		dpVals[i] = dp.DpVal
+		tsids[i] = dp.Tsid
+	}
+
+	for _, ts := range timestamps {
+		err := binary.Write(rawBlockBuf, binary.LittleEndian, ts)
+		if err != nil {
+			log.Errorf("EncodeWALBlock: failed to write timestamp: %v", err)
+			return err
+		}
+	}
+
+	for _, val := range dpVals {
+		err := binary.Write(rawBlockBuf, binary.LittleEndian, val)
+		if err != nil {
+			log.Errorf("EncodeWALBlock: failed to write dpVal: %v", err)
+			return err
+		}
+	}
+
+	for _, id := range tsids {
+		err := binary.Write(rawBlockBuf, binary.LittleEndian, id)
+		if err != nil {
+			log.Errorf("EncodeWALBlock: failed to write tsid: %v", err)
+			return err
+		}
+	}
+
+	compressedData, err := compressDataWithZSTD(rawBlockBuf.Bytes())
+	if err != nil {
+		log.Errorf("EncodeWALBlock: zstd compression failed: %v", err)
+		return err
+	}
+
+	_, err = compressedBlockBuf.Write(compressedData)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write compressed data to buffer: %v", err)
+		return err
+	}
+	return nil
+}
+
+func compressDataWithZSTD(data []byte) ([]byte, error) {
+	encoder, err := zstd.NewWriter(nil)
+	if err != nil {
+		log.Errorf("zstdCompressBytes: failed to create zstd encoder: %v", err)
+		return nil, err
+	}
+	defer encoder.Close()
+	return encoder.EncodeAll(data, nil), nil
```

Comment:
make this a class function, and the second param you pass to EncodeAll, pass a class variable encodeBuf, take this buf from the buffer pool and keep reusing .
see how the segwriter is doing it here : 
https://github.com/siglens/siglens/blob/4ea5e452855dbc720a20289a01a6bfdb1c540677/pkg/segment/writer/segwriter.go#L974